### PR TITLE
memsup: configurable alarm src

### PR DESCRIPTION
**Pull request by qzhuyan**
memsup: configurable alarm src
    
    In Linux,
    
    For the runtime system that is managed by Linux cgroups such as lxc and
    docker container, erlang VM should read meminfo from procfs: /proc/meminfo to
    get correct view of memory usage and the limit.
    
    Currently 'memsup' supports read from procfs via memsup:get_system_memory_data/0.
    However the memsup alarm still reads via sysconf call.
    
    This commit adds a config switch to make memsup get alarm input from 'procfs'
    or 'sysconf call' via application env: 'meminfo_src'.

### CODE REVIEW

File: lib/os_mon/src/memsup.erl [hunk 1]
```diff
@@ -344,16 +344,22 @@ handle_cast(_Msg, State) ->
...
+	{undefined, undefined} ->
 	    WDTimer = erlang:send_after(State#state.helper_timeout,
 					self(),
 					reg_collection_timeout),
 	    if
 		State#state.port_mode ->
-		    State#state.pid ! {self(), collect_sys},
-		    {noreply, State#state{wd_timer=WDTimer,
-					  pending=[reg]}};
+		    State#state.pid ! {self(), MemInfoSrc},
```
Comment (qzhuyan) [inline lib/os_mon/src/memsup.erl]: move port_mode check up.

File: lib/os_mon/src/memsup.erl [hunk 2]
```diff
@@ -493,7 +485,17 @@ handle_info({collected_ext_sys, SysMemUsage}, State) ->
     %% Send the reply to all waiting clients, preserving time order
     reply(State#state.ext_pending, undef, SysMemUsage),
 
-    {noreply, State#state{ext_wd_timer=undefined, ext_pending=[]}};
+    Total = proplists:get_value(system_total_memory, SysMemUsage),
```
Comment (qzhuyan) [inline lib/os_mon/src/memsup.erl:487]: we should also start a timer if  `reg` in the wait list.

File: lib/os_mon/test/memsup_SUITE.erl [hunk 3]
```diff
@@ -330,12 +347,12 @@ alarm2(_Config, _SysUsage) ->
     ok = memsup:set_check_interval(60),
 
     %% Check data and thresholds
-    {Total, Alloc, undefined} = memsup:get_memory_data(),
+    {_Total, _Alloc, undefined} = memsup:get_memory_data(),
```
Comment (qzhuyan) [inline lib/os_mon/test/memsup_SUITE.erl:350]: remove this line

**Comment by qzhuyan** [comment]
This PR is not urgent to merge, but there is another alternative in app `lc` see PR: https://github.com/qzhuyan/lc/pull/6

**Review by ieQu1 — APPROVED** [review]

File: lib/os_mon/test/memsup_SUITE.erl [hunk 4]
```diff
@@ -205,14 +226,11 @@ alarm1(_Config, SysUsage) ->
...
-                              if
-                                  Value > 0.99 -> 0.99;
-                                  true -> Value
-                              end;
-                          not SysP -> 0.9*SysUsage
+    NewSysThreshold = case SysP of
+                          true ->
+                              1.0;
+                          false ->
+                              0.0
```
Comment (thalesmg) [inline lib/os_mon/test/memsup_SUITE.erl:233]: Just confirming: these limits seem quite different from the previous ones.  Is this change intentional?  Maybe it was only for ease of testing?

File: lib/os_mon/src/memsup.erl [hunk 5]
```diff
@@ -843,6 +850,21 @@ clear_alarms() ->
...
+        Bytes > Threshold ->
+            set_alarm(process_memory_high_watermark, Pid);
+
+        true ->
+            clear_alarm(process_memory_high_watermark)
+    end,
+    State#state{mem_usage={Alloc, Total},
+                worst_mem_user={Pid, Bytes}};
+maybe_check_process_mem(State, _) ->
+    State.
```
Comment (thalesmg) [inline lib/os_mon/src/memsup.erl:866]: Maybe should set `mem_usage` as [before](https://github.com/emqx/otp/pull/21/files#diff-39f19df8821650991a8e7b166f5b8754b2103186f31d6c4b5eb9ec7de1207bc7L416)?

```suggestion
    State#state{mem_usage={Alloc, Total}}.
```

File: lib/os_mon/src/memsup.erl [hunk 6]
```diff
@@ -492,8 +484,23 @@ handle_info({collected_ext_sys, SysMemUsage}, State) ->
...
+                    set_alarm(system_memory_high_watermark, []);
+                true ->
+                    clear_alarm(system_memory_high_watermark)
+            end,
+            erlang:send_after(State#state.timeout, self(), time_to_collect),
+            ok;
+        false ->
+            skip
+    end,
+    NewState = maybe_check_process_mem(State, {Alloc, Total}),
```
Comment (thalesmg) [inline lib/os_mon/src/memsup.erl]: Q: since the `true` branch just above is doing something very similar to `maybe_check_process_mem`, is there any problem in possibly setting the alarm twice?

Comment (qzhuyan) [reply lib/os_mon/src/memsup.erl]: But they are two types of alarms, `system_memory_high_watermark` and `process_memory_high_watermark`

Comment (thalesmg) [reply lib/os_mon/src/memsup.erl]: Oops, you're right.  Ignore me.  :speak_no_evil:

File: lib/os_mon/src/memsup.erl (hunk 5 above)
Comment (qzhuyan) [reply lib/os_mon/src/memsup.erl:866]: Nice catch. 
I double checked that `mem_usage` is used by get_memory_data/1 which is different than `get_system_memory_data` so I have to update the `mem_usage` in the caller instead.

File: lib/os_mon/test/memsup_SUITE.erl (hunk 4 above)
Comment (qzhuyan) [reply lib/os_mon/test/memsup_SUITE.erl:233]: There is a logic error in the old code that if the system mem is already >99%, the test case will fail.

File: lib/os_mon/src/memsup.erl [hunk 7]
```diff
@@ -344,28 +344,39 @@ handle_cast(_Msg, State) ->
...
-	    if
-		State#state.port_mode ->
-		    State#state.pid ! {self(), collect_sys},
-		    {noreply, State#state{wd_timer=WDTimer,
-					  pending=[reg]}};
-		true ->
-		    OS = State#state.os,
-		    Self = self(),
-		    Pid = spawn_link(fun() ->
+					self(), reg_collection_timeout),
```
Comment (qzhuyan) [inline lib/os_mon/src/memsup.erl:352]: It should be `ext_collection_timeout` for collect_ext_sys

**Comment by qzhuyan** [comment]
PR to upstream
